wip - adding IPBlock and Egress rule support

nelljerram opened the pull request:
## Description
Extend KDD's k8s->Calico conversion so as to handle the following possible new elements in NetworkPolicy:
- IPBlock
- EgressRules.

## Todos
- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note
<!-- Writing a release note:
- By default, no release note action is required.
- If you're unsure whether or not your PR needs a note, ask your reviewer for guidance.
- If this PR requires a release note, update the block below to include a concise note describing
  the change and any important impacts this PR may have.
-->

```release-note
None required
```

Commit:
Incorporate k8s.io/networking policy into libcalico-go

Commit:
point REST client to v1beta1

Commit:
use new RESTClient, all FVs pass

Commit:
rename client

Commit:
fixup comments

Commit:
couple more comments fixed

Commit:
cleaning up debug logging

Commit:
adding FV that verifies we can write / read egress rules

Commit:
Use 127.0.0.1 when bootstrapping content into API server

For some reason, 'localhost' resolves to some 92.x.y.z address for me,
and then the crds.yaml step fails.

Commit:
Trace update types as strings, as numbers are harder

Commit:
Expect New update in "should handle a basic Pod" test

Commit:
KDD: Convert NetworkPolicy IP blocks and egress rules

caseydavenport commented on lib/backend/k8s/conversion.go:
```diff
@@ -230,7 +230,21 @@ func (c Converter) NetworkPolicyToPolicy(np *extensions.NetworkPolicy) (*model.K
+		outboundRules = append(outboundRules, c.k8sRuleToCalico(r.To, r.Ports, np.ObjectMeta.Namespace)...)
+	}
+
+	// Calculate Types setting.
```
I think we need to use the exact values specified in `np.Spec.PolicyTypes` rather than calculating it here ourselves.

nelljerram replied:
D'oh yes, of course!

caseydavenport commented on lib/backend/k8s/conversion.go:
```diff
@@ -382,18 +404,42 @@ func (c Converter) k8sProtocolToCalico(protocol *extensions.Protocol) *numorstri
+		_, ipNet, err := cnet.ParseCIDR(peer.IPBlock.CIDR)
+		if err != nil {
+			log.WithField("cidr", peer.IPBlock.CIDR).WithError(err).Error("Failed to parse CIDR")
+			goto parseError
```
Does this `goto` buy us anything over just returning here?  Seems like that would be easier to follow.

nelljerram replied:
Well then I have to repeat the 'null' return tuple (i.e. `"", nil, nil`)  I thought it could be worth a goto to avoid that.

nelljerram replied:
I guess I could set up the null return values at the top of the func, and then just return here...

caseydavenport reviewed (changes requested):
Just a couple of comments, but otherwise looking good.

caseydavenport commented on lib/backend/k8s/conversion.go:
```diff
@@ -354,23 +370,29 @@ func (c Converter) k8sIngressRuleToCalico(r extensions.NetworkPolicyIngressRule,
 		Action:      "allow",
 		Protocol:    protocol,
-		SrcSelector: srcSelector,
+		SrcSelector: selector,
```
Hm, I don't think this works actually, since this function gets called for both ingress and egress rule generation.

Don't we need to set `SrcSelector` / `DstSelector` accordingly?

Same for the `Nets` fields.

nelljerram replied:
Indeed, good catch.

nelljerram replied:
But interestingly, ports are always DstPorts, in both ingress and egress rules.

caseydavenport commented on lib/backend/k8s/conversion.go:
```diff
@@ -354,23 +370,29 @@ func (c Converter) k8sIngressRuleToCalico(r extensions.NetworkPolicyIngressRule,
 	if peer != nil {
 		// Peer information available.
-		srcSelector = c.k8sPeerToCalicoSelector(*peer, ns)
+		selector, nets, notNets = c.k8sPeerToCalicoSrcFields(*peer, ns)
```
This also needs a rename to remove the `Src` bit

Commit:
Review markups

Commit:
Add explicit PolicyTypes to test cases with no rules

These cases all expect Types ['ingress'] after conversion to Calico;
we need to add an explicit PolicyTypeIngress to get that.
